### PR TITLE
SLING-13171 use Apache Johnzon for JSON parsing

### PR DESCRIPTION
**Pull request by joerghoh**
Makes it easier to also handle invalid JSON

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -111,18 +111,24 @@
...
+            <groupId>org.apache.geronimo.specs</groupId>
+            <artifactId>geronimo-json_1.1_spec</artifactId>
+            <version>1.5</version>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.johnzon</groupId>
+            <artifactId>johnzon-core</artifactId>
+            <version>${johnzon.version}</version>
+            <scope>provided</scope>
```
Comment (kwin) [inline pom.xml]: Why this dependency with this scope? At compile time only the API should be necessary

File: src/main/java/org/apache/sling/i18n/impl/JcrResourceBundle.java [hunk 2]
```diff
@@ -245,62 +251,29 @@ private Map<String, Object> loadFully(
...
+                try (InputStreamReader reader = new InputStreamReader(input, charset);
+                        JsonParser parser = JsonProvider.provider().createParser(reader)) {
+                    String key = null;
+                    while (parser.hasNext()) {
+                        final Event event = parser.next();
+                        if (event == Event.KEY_NAME) {
+                            key = parser.getString();
+                        } else if (event == Event.VALUE_STRING && key != null) {
+                            targetDictionary.put(key, parser.getString());
+                        }
```
Comment (kwin) [inline src/main/java/org/apache/sling/i18n/impl/JcrResourceBundle.java:272]: I think we should log unsupported elements.

**Review by kwin — CHANGES_REQUESTED** [review]

**Comment by henrykuijpers** [comment]
Not sure what the behavior is when Johnson is parsing a JSON that has duplicate keys, but Adobe Experience Manager recently (I believe 6.6.2) introduced duplicate keys in the i18n json files. 

That could start failing with this change, which is okay of course, because a JSON should not contain duplicate keys.

**Comment by kwin** [comment]
> Not sure what the behavior is when Johnson is parsing a JSON that has duplicate keys, but Adobe Experience Manager recently (I believe 6.6.2) introduced duplicate keys in the i18n json files. 
> 
> 
> 
> That could start failing with this change, which is okay of course, because a JSON should not contain duplicate keys. 

That can be configured in Johnzon via https://johnzon.apache.org/johnzon-core/apidocs/org/apache/johnzon/core/RejectDuplicateKeysMode.html

File: pom.xml (hunk 1 above)
Comment (joerghoh) [reply pom.xml]: right, this is not required.

File: src/main/java/org/apache/sling/i18n/impl/JcrResourceBundle.java (hunk 2 above)
Comment (joerghoh) [reply src/main/java/org/apache/sling/i18n/impl/JcrResourceBundle.java:272]: we have a lot of events, which this code does not handle; and I am not sure if logging them makes sense.

**Comment by joerghoh** [comment]
I would opt to stick to the "last definition wins", which existed prior to this change, and which also now is implemented by now.

File: src/main/java/org/apache/sling/i18n/impl/JcrResourceBundle.java (hunk 2 above)
Comment (joerghoh) [reply src/main/java/org/apache/sling/i18n/impl/JcrResourceBundle.java:272]: (also the existing code behaves in that particular way, and it ignores silently elements, which are not used)

Comment (kwin) [reply src/main/java/org/apache/sling/i18n/impl/JcrResourceBundle.java:272]: Ok, this can be improved separately.

File: pom.xml [hunk 3]
```diff
@@ -111,18 +110,18 @@
...
             <artifactId>org.apache.sling.api</artifactId>
             <version>3.0.0</version>
             <scope>provided</scope>
         </dependency>
+        <dependency>
+            <groupId>org.apache.geronimo.specs</groupId>
+            <artifactId>geronimo-json_1.1_spec</artifactId>
+            <version>1.5</version>
+            <scope>provided</scope>
+        </dependency>
```
Comment (kwin) [inline pom.xml:124]: I would add a direct test dependency on johnzon-core, right now it is only transitively included via sling-mock.

**Review by kwin — APPROVED** [review]
LGTM, just one minor nit.

File: src/test/java/org/apache/sling/i18n/impl/JcrResourceBundleTest.java [hunk 4]
```diff
@@ -447,4 +452,40 @@ public void test_json_dictionary() throws Exception {
...
+    private Resource mockJsonResource(String json, String encoding) {
+        Resource resource = Mockito.mock(Resource.class);
+        ResourceMetadata metadata = Mockito.mock(ResourceMetadata.class);
+        Mockito.when(resource.getPath()).thenReturn("/apps/i18n/de.json");
+        Mockito.when(resource.getResourceMetadata()).thenReturn(metadata);
+        Mockito.when(metadata.getCharacterEncoding()).thenReturn(encoding);
+        Mockito.when(resource.adaptTo(InputStream.class))
+                .thenReturn(new ByteArrayInputStream(json.getBytes(StandardCharsets.UTF_8)));
+        return resource;
+    }
```
Comment (kwin) [inline src/test/java/org/apache/sling/i18n/impl/JcrResourceBundleTest.java:477]: Adding a test for duplicate keys would be nice to have.

Comment (kwin) [reply src/test/java/org/apache/sling/i18n/impl/JcrResourceBundleTest.java:477]: Added via https://github.com/apache/sling-org-apache-sling-i18n/pull/23.

File: pom.xml (hunk 3 above)
Comment (joerghoh) [reply pom.xml:124]: done

File: src/test/java/org/apache/sling/i18n/impl/JcrResourceBundleTest.java (hunk 4 above)
Comment (joerghoh) [reply src/test/java/org/apache/sling/i18n/impl/JcrResourceBundleTest.java:477]: thanks!
